File: src/LongestCommonPrefix.java
/**
 * User: guoxc
 * Date: 15-1-7
 * Time: 17:31
 */
public class LongestCommonPrefix {
    public String longestCommonPrefix(String[] strs){
        if(strs.length == 0){
            return "";
        }
        int[] lens = new int[strs.length];
        int min_len = strs[0].length();
        int min_index = 0;
        for(int i = 0; i < strs.length; i += 1){
            lens[i] = strs[i].length();
            if(lens[i] < min_len){
                min_len = lens[i];
                min_index = i;
            }
        }
        int r = 0;
        for(int i = 0; i < min_len; i += 1){
            char c = strs[min_index].charAt(i);
            boolean flag = false;
            for(int j = 0; j < strs.length; j += 1){
                if(strs[j].charAt(i) != c){
                    flag = true;
                    break;
                }
            }
            if(!flag){
                r += 1;
            }else{
                break;
            }
        }
        return strs[min_index].substring(0,r);
    }
    public static void test(){
        System.out.println(new LongestCommonPrefix().longestCommonPrefix(
            new String[]{"abcd","abcx"}
        ));
    }
}

File: cpp/longest_common_prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        int n = strs.size();
        if(n == 0) return "";
        int min_L = strs[0].length();
        int min_i = 0;
        for(int i = 1; i < n; i += 1){
            if(strs[i].length() < min_L){
                min_L = strs[i].length();
                min_i = i;
            }
        }
        int l = 0;
        for(int i = 0; i < strs[min_i].length(); i += 1){
            char c = strs[min_i][i];
            for(int j = 0; j < n; j += 1){
                if(j == min_i) continue;
                if(strs[j][i] != c){
                    return strs[min_i].substr(0, l);
                }
            }
            l += 1;
        }
        return strs[min_i].substr(0, l);
    }
};
